tests/test_ruby_server.py

```diff
@@ -0,0 +1,193 @@
+          'string' => {
+            'action.account_id' => @action_account_id.to_s,
+            'action.controller' => @action_controller.to_s,
+            'action.method' => @action_method.to_s,
+            'action.path' => @action_path.to_s,
+            'action.status_code' => @action_status_code.to_s,
+            'action.user_id' => @action_user_id.to_s,
+          },
+        },
+      }
+      serialized_event_payload = event_payload.to_json
+      # This is the message structure that Decoder expects: https://github.com/mozilla/gcp-ingestion/pull/2400.
+      ping = {
+        'document_namespace' => @application_id,
+        'document_type' => 'mastodon-action',
+        'document_version' => '1',
+        'document_id' => SecureRandom.uuid,
+        'user_agent' => @user_agent,
+        'ip_address' => @ip_address,
+        'payload' => serialized_event_payload,
+      }
+      # This structure is similar to how FxA currently logs with mozlog: https://github.com/mozilla/fxa/blob/4c5c702a7fcbf6f8c6b1f175e9172cdd21471eac/packages/fxa-auth-server/lib/log.js#L289
+      logger = Logger.new($stdout)
+      logger_name = 'mastodon-action-glean'
+      logger.formatter = proc do |severity, datetime, _progname, msg|
+        date_format = datetime.to_i
+        "#{JSON.dump(Timestamp: date_format.to_s, Logger: logger_name.to_s, Type: GLEAN_EVENT_MOZLOG_TYPE.to_s, Severity: severity.ljust(5).to_s, Pid: Process.pid.to_s, Fields: msg)}\n"
+      end
+      logger.info(ping)
+    end
+  end
+end
+
+# use the following examples for adding the loggers to your controller
+# class ApplicationController < ActionController::Base
+    # # add glean server side log for controller calls
+    # around_action :emit_server_side_events
+
+# ...
+
+# private
+# def emit_server_side_events
+  # yield
+# ensure
+  # GleanHelper::MastodonActionServerEvent.new(
+    # application_id:'ruby app name',
+    # app_display_version:'ruby app name as `X.X.X`',
+    # app_channel:'environment for exampe, `production` or `development`',
+    # user_agent:'string or expression',
+    # ip_address:'string or expression',
+    # action_account_id:'string or expression',
+    # action_controller:'string or expression',
+    # action_method:'string or expression',
+    # action_path:'string or expression',
+    # action_status_code:'string or expression',
+    # action_user_id:'string or expression',
+  # ).record
+# end
+# end"""
+        assert content_1.strip().replace(" ", "").replace("\n", "").replace(
+            "\\n", ""
+        ) in content.strip().replace(" ", "").replace("\n", "").replace("\\n", "")
+        assert content_2.strip().replace(" ", "").replace("\n", "").replace(
+            "\\n", ""
+        ) in content.strip().replace(" ", "").replace("\n", "").replace("\\n", "")
+        assert content_3.strip().replace(" ", "").replace("\n", "").replace(
+            "\\n", ""
+        ) in content.strip().replace(" ", "").replace("\n", "").replace("\\n", "")
```

Comment:
That's a whole lot of whitespace replacing and it's completely unclear why that is happening. Why can't we compare the content exactly?

Reply:
this is just an artifact of splitting up the comparison text inline.  Lets see how the file comparison works.